### DISCUSSION TITLE
blog/inheritance-angular-inject

### DISCUSSION THREAD

**Discussion by giscus** [nartc/tnc-scully#35] [Announcements]
# blog/inheritance-angular-inject

Angular 14 comes with a subtle, yet powerful, change to the Dependency Injection system that might be a game-changer for many Angular developers

https://nartc.me/blog/inheritance-angular-inject

**Comment by stephanietuerk** [comment]
Thanks for this write up -- not having to worry about the chain of dependencies when extending is indeed really nice! 

I am wondering, however, if you had any thoughts on the implications of this for unit testing -- specifically, how one could stub out services / whatever is injected in the extended/base class?

**Reply by stephanietuerk** [reply]
Ah! I think I've solved this. Provided that you use the TestBed, you can provide dependencies -- or stubs -- in the same way as you usually would for testing. (In my case I was not using the TestBed initially as my concrete class was a Directive, not a component.)

**Reply by raunaky** [reply]
stuck with the same situation where i need to write unit test for my concrete component (subclass) extending an abstract class (base class) which has dependencies being injected with the inject() method

**Reply by stephanietuerk** [reply]
Did you try providing the dependencies through the TestBed providers array? Whether you use `inject` or the traditional way of providing the dependency through the constructor doesn't matter from the perspective of the test. In the test file, you still need to provide all dependencies for both the base/abstract class and for the concrete component.

**Reply by raunaky** [reply]
I provided all the dependencies in the providers:[]
But still facing the error Nullinjection : R3jnjectionError(DynamicTestModule[StoreFeatureModule-> ReducerManager ->ReducerManager)

**Reply by stephanietuerk** [reply]
It's hard for me to help any further without seeing your code. If you are still stuck, I'd recommend asking in the Angular Discord.

**Comment by PhiLhoSoft** [comment]
Before Angular 14, the abstract class was given only `injector: Injector`, and then, in the constructor I was doing `this.serviceOne = injector.inject(ServiceOne)`, etc. So the sub-class only had to pass the injector to the parent. Thus I had some flexibility in extending the base class without rewriting all the sub-classes.
Since then, I replaced all this with `inject()`, much better!
